contracts/automation/andromeda-task-balancer/src/contract.rs

```diff
@@ -0,0 +1,428 @@
+
+    let storage_address = loop {
+        let has_process = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: contract_addr[num as usize].clone(),
+            msg: to_binary(&StorageQueryMsg::HasProcess {
+                process: deps.api.addr_validate(&process)?,
+            })?,
+        }))?;
+        if has_process {
+            break &contract_addr[num as usize];
+        }
+        num += 1;
+
+        if num as usize >= (contract_addr.len()) {
+            return Err(ContractError::ProcessNotFound {});
+        }
+    };
+    // Mark that storage contract by adding it to the UP_NEXT vector, which allows us to fill that empty space in the future
+    let up_next = UP_NEXT.may_load(deps.storage)?;
+
+    if let Some(mut up_next) = up_next {
+        up_next.push(storage_address.clone());
+        UP_NEXT.save(deps.storage, &up_next)?;
+    } else {
+        let up_next = vec![storage_address.clone()];
+        UP_NEXT.save(deps.storage, &up_next)?;
+    }
+
+    Ok(Response::new()
+        .add_submessage(SubMsg::new(CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: storage_address.clone(),
+            msg: to_binary(&StorageExecuteMsg::Remove {
+                process: process.to_string(),
+            })?,
+            funds: vec![],
+        })))
+        .add_attribute("action", "removed_process")
+        .add_attribute("process", process)
+        .add_attribute("up_next", storage_address))
+}
+
+fn add_process(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    process: String,
+) -> Result<Response, ContractError> {
+    // Not anyone should be allowed to add tasks to tree
+    let state = STATE.load(deps.storage)?;
+    ensure!(info.sender == state.admin, ContractError::Unauthorized {});
+
+    // Validate the process's address
+    let process = deps.api.addr_validate(&process)?;
+
+    // In case no storage contracts have been instantiated yet
+    if state.contracts == Uint128::zero() {
+        STATE.update(deps.storage, |mut state| -> Result<_, ContractError> {
+            state.contracts += Uint128::from(1u64);
+            Ok(state)
+        })?;
+        let state = STATE.load(deps.storage)?;
+        let msg = CosmosMsg::Wasm(WasmMsg::Instantiate {
+            admin: None,
+            code_id: state.storage_code_id,
+            msg: to_binary(&StorageInstantiateMsg {
+                task_balancer: env.contract.address,
+                process,
+                max_processes: state.max,
+            })?,
+            funds: vec![],
+            label: "storage".to_string(),
+        });
+        return Ok(Response::new()
+            .add_attribute("action", "try_add")
+            .add_submessage(SubMsg::reply_on_success(msg, 1)));
+    }
+
+    // Check if there are any earlier storage contracts with free space
+    let up_next = UP_NEXT.may_load(deps.storage)?;
+
+    if let Some(mut up_next) = up_next {
+        if up_next.is_empty() {
```

Comment:
Is there a reason for this to be if/else? Can we have `if !up_next.is_empty()` instead?